Refactor(ui): tweak ui logic

YvonTre opened the pull request:
## Video State Changing Cycle

- All changable state(except for `currentTime`) must go through Vuex first, then to controller component and finally to base component. Here are some examples:

### When user wants to play/pause video

1. User clicks on the screen/press the space key.
2. Trigger global event `toggle-playback`.
3. `VideoCanvas`(As `BaseVideoPlayer`'s controller component) dispatch `play` or `pause` Vuex action.
4. Vuex state changes.
5. `VideoCanvas`'s Vuex `paused` getters changes.
6. `BaseVideoPlayer`'s `paused` props changed.
7. Video paused/played.

### When user wants to seek to some time

1. User clicks on the progress bar/press the left/right key.
2. Event listeners emit global seek event.
3. `VideoCanvas` change `BaseVideoPlayer`'s `currentTime` props.
4. `BaseVideoPlayer` change video's currentTime.
5. `BaseVideoPlayer` commit Vuex's `currentTime` mutations.

## Refactor Vuex

- Put mutation types and action types in `mutation-types.js` and `action-types.js`.
- Do not commit mutations directly, dispatch actions instead.
- Do validations in actions.
- Do not use state directluy components, use getters instead.

## Update `.eslintrc.js`

- Change complexity from 'warn' to 'error'.

## Clean up tests

- Delete all unnecessary `console.log`.
- Use sinon's fakeTimers when testing `requestAnimationFrame`.

### Fix Error Logs

#### `Error: [vuex] Do not mutate vuex store state outside mutation handlers.`

- Avoid mutating Vuex state directly.
- Avoid committing Vuex mutations. Use fake actions/mutations instead.

#### `[vuex] unknown getter: xxxxxx`

- Add lacked getters in some tests.

#### `[vuex] unknown action type: XXXXXX`

- Add lacked actions in some tests.

#### `[Vue warn]: Invalid prop: custom validator check failed for prop "defaultEvents".` && `[Vue warn]: Missing required prop: "defaultEvents"`

- Check propsData carefully when mounting/shallowMounting components.
- Avoid passing unnecessary props around components.

#### `[Vue warn]: Invalid default value for prop "xxxxxx": Props with type Object/Array must use a factory function to return the default value.`

- Use function like `() => {}/[];` as default props value for Object/Array.

#### `[Vue warn]: Avoid adding reactive properties to a Vue instance or its root $data at runtime - declare it upfront in the data option.`

- Avoid to use `wrapper.setData` on non-existed data properties.
- When clear unnecessary data properties, check whether unit tests have some legacy setData methods.

#### `TypeError: Cannot read property 'xxxxxx' of undefined/null`

- Improve function's robustness and better error handling capabilities.

## Rewrite TimeProgressBar

- Refactor a new TheProgressBar based on old TimeProgressBar.
- Use computed properties as much as possible.
- Use CSS to control style as much as possitble.
- Simplify hovered and played display logic.

## To-do

- Draft a unit-test style guide and FAQ.
- Refactor BaseSubtitle and SubtitleControl and their tests.
- Remove console.log completely.

Commit:
refactor(video,control): update volume and playbackRate logic

Commit:
refactor(vuex,video): implement all of video attributes

Commit:
refactor(vuex,src): adjust video src's source

Commit:
refactor(base,video): refactor BaseVideoPlayer

1. Extract all attributes can be changed as props, remove defaultOptions
2. Replace some initial functions as more robust generic methods.
3. Change currentTime update logic.

Commit:
refactor(video): change VideoCanvas and vuex accordingly

Commit:
refactor(ui): tweak ui logics to change with Vuex getters

Commit:
Resolve all conflicts.

Commit:
refactor(vuex,validators): remove all validators

Commit:
refactor(video,complexity): reduce complexity and update rule to error

Commit:
test(base,video): fix BaseVideoPlayer tests

Commit:
refactor(play): refactor PlayButton and fix its unit tests

Commit:
refacor(base,video): add emitEvents and fix tests

Commit:
Resolve all conflicts.

Commit:
fix(seek): fix seek functionality

ipy commented on src/renderer/components/PlayingView.vue:
```diff
@@ -16,14 +16,7 @@ export default {
-    uri() {
-      return this.$store.getters.convertedSrcOfVideo;
-    },
+    this.$electron.remote.getCurrentWindow().setMinimumSize(320, 180);
```
这个调整原因是？

ipy commented on src/renderer/components/PlayingView/TimeProgressBar.vue:
```diff
@@ -255,13 +243,14 @@ export default {
-      return (this.$store.state.PlaybackState.CurrentTime
-        / this.$store.state.PlaybackState.Duration) * progressBarWidth;
+      return (this.currentTime
+        / this.duration) * progressBarWidth;
```
类似这种应该可以合成一行了，之前换行应该是为了不超过单行字符数限制

ipy commented on src/renderer/store/mutation-types.js:
```diff
@@ -1,7 +1,34 @@
 export const Video = {
```
除了这两个文件应该没有其他文件名使用连字符，建议统一格式

Commit:
refactor(vuex): rename some vuex getters for better readability

Commit:
fix(bug,vuex): fix vuex-related problems

Commit:
chore(playlist): rename vuex module from PlaybackState to Playlsit

Commit:
chore(vuex,module): rename two modules

Commit:
test(vuex): rename import statements

Commit:
chore(rm,mv): remove and move files

Commit:
chore(vuex): rename some vuex modules

Commit:
test(ui,timecodes):refactor TheTimeCodes test

Commit:
fix(ui,thumbnail): fix ThePreviewThumbnail not working properly

Commit:
chore(vuex,names): rename mutation-types and action-types

Commit:
chore(vuex): change mutations names

Commit:
refactor(vuex): add meta info state, mutations and helper functions

Commit:
refactor(video,info): commit video meta info to vuex and remove bus event

Commit:
refactor(vuex): refactor Window module and VideoCanvas

Commit:
test(refactor): refactor tests

Commit:
refactor(video): use new calculateWindowSize function

Commit:
refactor(video): add calculateWindowPosition and getWindowRect

Commit:
refactor(video): do some clean-up

Commit:
test(video): add test for VideoCanvas

Commit:
test(fix): fix some test problems

Commit:
refactor(ui,progress): fix timeProgressBar's ratio data

Commit:
fix(test): fix BaseVideoPlayer's requestAnimationFrame test

Use sinon.useFakeTimers to avoid problems caused.

Commit:
test(windowSizeHelper): add timer.restore() to avoid problems

Commit:
test(fix): fix all test's problems

1. Prevent open file diagoue in tests.
2. Prevent direct changes to vuex state in tests.
3. Use fake timers to fix requestAnimationFrame tests.
4. Use fake actions to prevent direct mutations in Vuex.
5. Delete all BaseSubtitle unit tests.
6. Remove unnecessary tests.
7. Fix Vuex getters undefined error log.

Commit:
refactor(ui): add new widget TheProgressBar

Commit:
feat(ui, progress): add document drag functionality

Commit:
feat(ui, progress): add more hovering scenarios support

Commit:
feat(ui, progress): add ThePreviewThumbnail support

Commit:
feat(ui, progress): add fakeButton support

Commit:
refactor(ui, progress): remove TimeProgressBar add TheProgressBar tests

Commit:
refactor(ui, progress): add TheProgressBar to TheVideoController

Commit:
test(fix): fix some tests and components

1. Replace invalid watcher.
2. Tempoarily disable some console.log.
3. Remove some setData for non-existed data.
4. Use beforeEach's shallowMount for all tests.

Commit:
Resolve all conflicts.

ipy commented on src/renderer/helpers/infoDB.js:
```diff
@@ -65,11 +65,10 @@ class InfoDB {
       const tx = db.transaction(schema, 'readwrite');
       tx.objectStore(schema).put(data);
-      return tx.complete.then(() => console.log('added'));
+      return tx.complete.then(() => 'added');
```
返回的这个字符串有用到么

YuRiJinX replied:
谭扬的pr已经改过了 @ipy

YuRiJinX commented on src/renderer/components/PlayingView/NextVideo.vue:
```diff
@@ -15,7 +15,7 @@
         <div class="top">
           <div class="duration">{{ duration }}</div>
-          <div class="title">&nbsp· {{ title }}</div>
+          <div class="title">&nbsp;· Next in Playing List</div>
```
这个改回去啊

Commit:
fix(ui): fix TheProgressBar problems and add unit tests

1. Add mouseleave data to avoid progress  bar disappear upon mouseup.
2. Use currentTime not roundedCurrentTime for better progress moving.
3. Cancel root element's opacity to avoid weird transition problem.

Commit:
refactor(ui): change opacity-based transition to background-color

Commit:
fix(nextvideo): fix next video problem

Commit:
style(ui): change progress bar div transition

Commit:
fix(style): fix JumpToEnd button opacity problems

Commit:
refactor(core, video): make currentTime an array and fix test

Commit:
fix(ui): fix thumbnail show conditions and left fake progres's radius

Commit:
fix(ui, progress): fix progress bar invoke problem

Commit:
Resolve all conflicts.

Commit:
fix(ui): fix leftFakeButton hovering problem

Commit:
fix(core, nextvideo): fix nextVideo functionality

Commit:
fix(core, next): fix duration problem

Commit:
fix(style): fix fakeLeftButton's white opacity

YvonTre committed:
Merge branch 'develop' into refactor(ui)-tweak-ui-logic

Commit:
Merge branch 'develop' into refactor(ui)-tweak-ui-logic

Commit:
ci(windows): fix require electron problem and video controller
